### PR TITLE
Semantic Segmentation model_fcn_resnet50 and model_fcn_resnet_101 

### LINKED ISSUE #198: [Semantic segmentation model] Please implement FCN model family
FCN is a semantic segmentation model family based on [Fully Convolutional Networks for Semantic Segmentation](https://arxiv.org/abs/1411.4038) paper

Please implement

    fcn_resnet50
    fcn_resnet101

:Note: those two models pretrained weights are already available in https://torch-cdn.mlverse.org/models/vision/v1/models/

### PR DESCRIPTION
**Pull request by koshtiakanksha**
Closes #198

### CODE REVIEW

File: R/models-fcn.R
Comment (cregouby) [inline R/models-fcn.R:1]: **todo** your download_and_cache() values L122 and L152 are erroneous and need to be fixed to the proper fcn models.
**suggestion** having model URLs associated with MD5 hash and size is the perfect enabler for some missing convenioent features (see https://github.com/mlverse/torchvision/blob/14d17bc5e77f773872d43074c28c7ad26fe8d897/R/models-vit.R#L32-L37 ) 
**improvement** it would be nice to add a MD5 check of the downloaded model weights
**improvement** it would be nice to inform the end-user of the size of the model to be downloaded.
**improvement** please strictly reuse existing `cli_inform()` messages so that we don't need a specific translation per model (see https://github.com/mlverse/torchvision/blob/14d17bc5e77f773872d43074c28c7ad26fe8d897/R/models-vit.R#L44 )

File: tests/testthat/test-models-fcn.R
Comment (cregouby) [inline tests/testthat/test-models-fcn.R:1]: **todo** please add a test with non-default `num-classes` 
**improvement** **non-blocking** you could change the batch size of **one** of the tests to 2 or 3.
**todo** **missing** please also test `model_fcn_resnet101`

File: R/models-fcn.R [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+#' @examples
+#' model <- model_fcn_resnet50(pretrained = FALSE)
+#' input <- torch::torch_randn(1, 3, 224, 224)
+#' output <- model(input)
+#' str(output)
+#'
+#' model <- model_fcn_resnet101(pretrained = FALSE, aux_loss = TRUE)
+#' input <- torch::torch_randn(1, 3, 224, 224)
+#' output <- model(input)
+#' str(output)
```
Comment (cregouby) [inline R/models-fcn.R:29]: **improvement** could we make the examples more effectively being a segmented image with a workflow going through `draw_segmentation_masks()` and `tensor_image_display()` ? (see https://docs.pytorch.org/vision/main/auto_examples/others/plot_visualization_utils.html#visualizing-segmentation-masks)

**Review by cregouby — CHANGES_REQUESTED** [review]
**praise** thanks for this ! This opens the door to a nice vignette on segmentation end-to-end.
**todo** some changes should be made

**Comment by koshtiakanksha** [comment]
All requested changes have been addressed, including updated URLs, MD5 checks, complete tests, and improved examples. VOC class names are documented for now and will be exposed via a function once the dataset is added.

**Comment by koshtiakanksha** [comment]
Hi @cregouby, I’m encountering a CI error related to corrupt model files, but I don’t see any errors when running locally. I verified the MD5 checksums using: tools::md5sum("path/to/file.pth")
These values match the ones I’ve included in fcn_torchscript_urls. Do you have any suggestions on how to fix this or ensure consistency across local and CI enviro

File: R/models-fcn.R [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+#' @family models
+#' @name model_fcn_resnet
+#' @rdname model_fcn_resnet
+#'
+#' @examples
+#' \dontrun{
+#' model <- model_fcn_resnet50(pretrained = TRUE)
+#' input <- torch::torch_randn(1, 3, 224, 224)
+#' output <- model(input)
+#' mask <- output$out[1]$argmax(dim = 1)$unsqueeze(1)$to(torch::torch_bool())
```
Comment (cregouby) [inline R/models-fcn.R:31]: **praise** example is now really nice and explicit !
**todo** do not use implicit partial slicing `output$out[1]` in general but rather explicit torch slicing `output$out[1,..]`. In here, you should prefer squeezing on slicing : `output$out$squeeze(1)`
**todo** argmax(dim = 1) will provide a tensor with integer values 1 to 21 (the class id) so you cannot turn it into boolean (or you will always get the constant `torch_full()` tensor. 
**suggestion** turn the mask extraction into 2+ lines with explicit names and leading code comment so the end-user has a chance to understand what needs to be done with output.

File: R/models-fcn.R [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+#' img <- input[1]$mul(255)$to(dtype = torch::torch_uint8())
```
Comment (cregouby) [inline R/models-fcn.R:32]: **improvement** this is unneeded, please remove, since `draw_segmentation_mask` accept x Tensor of shape (C x H x W) and dtype uint8 or dtype float. In case of dtype float, values are assumed to be in range 
[0,1]

**Comment by koshtiakanksha** [comment]
@cregouby This is how it looks after normalizing the input. I have included it in the example for documentation. 

> norm_mean <- c(0.485, 0.456, 0.406)
> norm_std  <- c(0.229, 0.224, 0.225)
> img <- magick::image_read("tests/testthat/assets/class/cat/cat.1.jpg")
> input <- transform_to_tensor(img)
> input <- transform_resize(input, c(520, 520))
> input <- transform_normalize(input, norm_mean, norm_std)
> input <- input$unsqueeze(1)
> model <- model_fcn_resnet101(pretrained = TRUE)
Model weights for <model_fcn_resnet101> (~207 MB) will be downloaded and
processed if not already available.
> model$eval()
> output <- model(input)
> mask_id <- output$out$argmax(dim = 2)
> mask_bool <- torch::torch_stack(lapply(0:20, function(x) mask_id[1, ..]$eq(x)), dim = 1)
> segmented <- draw_segmentation_masks(input$squeeze(1), mask_bool)
> tensor_image_browse(segmented)

filef0fc48524188

**Review by cregouby — CHANGES_REQUESTED** [review]

**Comment by cregouby** [comment]
Hello @koshtiakanksha,

It looks like the new CI error is not related to the code in the PR but rather to something else on the CI machine, like a disk full or Out-of-memory. I'll have a deeper look tomorrow...

**Comment by cregouby** [comment]
Hello @koshtiakanksha,
I've fixed most of the R CMD check and tests errors. 

One test is still failing that requires a specific error management : 
```
── Error (test-models-fcn.R:35:3): model_fcn_resnet50 loads pretrained weights ───────────────────────────────────────────────────────────────────────────────────
<value_error/rlang_error/error/condition>
Error in `value_error("Could not find {key} in the state_dict.")`: Impossible de trouver classifier.4.weight dans ‘state_dict’.
Backtrace:
    ▆
 1. └─torchvision::model_fcn_resnet50(pretrained = TRUE, num_classes = 12) at test-models-fcn.R:35:3
 2.   └─model$load_state_dict(state_dict, strict = strict_loading) at torchvision/R/models-fcn.R:200:5
 3.     └─torch (local) load(self, state_dict)
 4.       └─torch (local) load(child, state_dict, prefix = paste0(prefix, nm, "."))
 5.         └─torch (local) load(child, state_dict, prefix = paste0(prefix, nm, "."))
 6.           └─module$.load_from_state_dict(state_dict, prefix)
 7.             └─torch:::value_error("Could not find {key} in the state_dict.")
 8.               └─rlang::abort(glue::glue(gettext(...)[[1]], .envir = env), class = "value_error")
[ FAIL 1 | WARN 0 | SKIP 0 | PASS 33 ]
```

For this test, the message is completley obscur to the end-user. 
**improvement** **blocking** This should be fix by the code or the end-user should be warn that pretrained=TRUE and num_classes cannot be configured together.
